### PR TITLE
Add type hints to optuna/trial

### PR DESCRIPTION
**Pull request by upura**
## Motivation
Please refer to the Issue #711 .

## Description of the changes
I added type hints into `optuna/trial/_base.py`, `optuna/trial/_fixed.py`, `optuna/trial/_state.py`, and `optuna/trial/_trial.py`.

https://github.com/optuna/optuna/issues/711#issuecomment-597088465

I encountered the importing problem of `optuna.study.Study` in updating `optuna/trial/_trial.py`, and I avoid this error by forward reference as suggested [here](https://github.com/optuna/optuna/issues/711#issuecomment-597088465).

## Local test
I ran two tests according to the [article](https://tech.preferred.jp/ja/blog/optuna-wants-your-pull-requests/).
```
$ pytest tests
$ circleci build --job checks
```

### CODE REVIEW

File: optuna/trial/_trial.py [hunk 1]
```diff
@@ -48,12 +44,7 @@ class Trial(BaseTrial):
 
     """
 
-    def __init__(
-        self,
-        study,  # type: Study
-        trial_id,  # type: int
-    ):
-        # type: (...) -> None
+    def __init__(self, study: "optuna.study.Study", trial_id: int,) -> None:
```
Comment (crcrpar) [inline optuna/trial/_trial.py:47]: This doesn't matter but for the consistency...
```suggestion
    def __init__(self, study: "optuna.study.Study", trial_id: int) -> None:
```

**Review by crcrpar — COMMENTED** [review]
Basically nice work 😃

**Review by HideakiImamura — APPROVED** [review]
Thank you for your contribution! LGTM except for @crcrpar's comment.

**Comment by upura** [comment]
Thanks for your reviews. Sorry for my carelessness.

**Review by crcrpar — APPROVED** [review]
You don't need to feel sorry at all. We are here to find and fix those things 😃

**Pull request merged**